doc/cps-model.md

```diff
@@ -0,0 +1,165 @@
+# How Groovy-CPS models CPS in Java
+In [CPS basics](cps-basics.md), we used the Java8 lambda notation to illustrate CPS translation/execution
+of the program as pseudo code.
+ Unfortunately we cannot actually model groovy-cps like that, because in Java we cannot write a function that
+doesn't return. That pseudo code will very quickly results in `StackOverflowError`.
+
+So in groovy-cps, we use `Next`, `Continuation`, and `Block` to model this.
+
+In pseudo-code, a continuation was a lambda of the form `(v)->{}`. In groovy-cps, `Continuation` interface
+represents a continuation. Here, instead of encapsulating the whole remaining program as single lambda,
+it encapsulates just the tiny logic of "receive the return value of a call and decide what to do next",
+then return "what to do next" as a `Next` object
+
+How does `Next` captures "what to do next"? `Next` is a pair of ...
+
+* `Block`, which represents a program (say, a function call)
+* `Continuation`, which represents how to process the value obtained by evaluating `Block` 
+
+Thus the way the execution goes is like:
+
+1. a `Block` starts execution, which (very quickly) returns a `Next`
+2. We get `Block` and `Continuation` from the `Next` previous step produced
+3. We start evaluating this new `Block` by giving it this new `Continuation`, which (very quickly) returns a `Next`
+4. We get `Block` and `Continuation` from the `Next` previous step produced
+3. We start evaluating this new `Block` by giving it this new `Continuation`, which (very quickly) returns a `Next`
+5. ... repeat ...
+
+Thus, the "interpreter" main loop is as follows. In groovy-cps, this code is actually in `Next.run()`:
+
+    Next n = <<program to execute>>
+    while(true) {
+        n = <<evaluate n.block with n.continuation>>
+    }
+    
+Ths main loop is a little bit like CPU instruction cycle. Fetch next instruction from [Program Counter](https://en.wikipedia.org/wiki/Program_counter),
+execute that one instruction, and repeat that forever. Each instruction is atomic.
+
+## Environment
+We need one more thing to run programs, which is "environments". If you mentally picture how Java program
```

Comment:
Might be worth saying environment == scope  (since "scope" may be an immediately recognizable term for many).